Modules/_sqlite/connection.c

```diff
@@ -321,10 +378,33 @@ free_callback_contexts(pysqlite_Connection *self)
     set_callback_context(&self->authorizer_ctx, NULL);
 }
 
+static void
+remove_callbacks(sqlite3 *db)
+{
+#ifdef HAVE_TRACE_V2
+    sqlite3_trace_v2(db, SQLITE_TRACE_STMT, 0, 0);
+#else
+    sqlite3_trace(db, 0, (void*)0);
+#endif
+    sqlite3_progress_handler(db, 0, 0, (void *)0);
+    (void)sqlite3_set_authorizer(db, NULL, NULL);
+}
+
 static void
 connection_close(pysqlite_Connection *self)
 {
     if (self->db) {
+        if (self->autocommit == AUTOCOMMIT_DISABLED &&
+            !sqlite3_get_autocommit(self->db))
```

Comment:
Regarding line 628 (`commit()`) and 662 (`rollback()`):

If you have autocommit disabled, you are in "transactional mode", so sqlite3 will keep a transaction open at any time, as per PEP 249. If the user messes this up by executing explicit transaction statements (for example `.execute("begin"`), a follow-up `.commit()` or `.rollback()` will generate an exception. This is fine; the user broke the contract.

For `close()`, it is a little bit different, because close may be called implicitly during connector tear-down (dealloc), which can happen in the middle of interpreter teardown. I'd rather not have `_pysqlite_seterror()` raise an error at that point.

I think this is fine, because the only two explicit "transaction control methods" are `.commit()` and `.rollback()`; `.close()` is not a "transaction control method".

I hope that was clarifying :)

Reply:
> If you have autocommit disabled, you are in "transactional mode", so sqlite3 will keep a transaction open at any time, as per PEP 249. If the user messes this up by executing explicit transaction statements (for example .execute("begin"), a follow-up .commit() or .rollback() will generate an exception. This is fine; the user broke the contract.

Actually the exception is raised by two `.execute('BEGIN')`, not by follow-up `.commit()` or `.rollback()`, for any commit modes:

```python
>>> import sqlite3
>>> cx = sqlite3.connect(':memory:', autocommit=False)
>>> cx.execute('BEGIN')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
sqlite3.OperationalError: cannot start a transaction within a transaction
>>> cx.commit()
>>> cx.commit()
>>> cx.rollback()
>>> cx.rollback()
>>> 
```

```python
>>> import sqlite3
>>> cx = sqlite3.connect(':memory:', autocommit=True)
>>> cx.execute('BEGIN')
<sqlite3.Cursor object at 0x10aabc170>
>>> cx.execute('BEGIN')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
sqlite3.OperationalError: cannot start a transaction within a transaction
>>> cx.commit()
>>> cx.commit()
>>> cx.rollback()
>>> cx.rollback()
>>> 
```

Reply:
> Is it the missing !sqlite3_get_autocommit(self->db) check for .commit() and .rollback() that made you expect they would raise an exception?

No. I should've used `COMMIT` iso. ´BEGIN` in my previous comment; my bad. This is what I hinted at:

```python
$ ./python.exe 
Python 3.12.0a1+ (heads/sqlite-autocommit-dirty:88e1420499, Nov 12 2022, 15:27:03) [Clang 13.1.6 (clang-1316.0.21.2.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sqlite3
>>> cx = sqlite3.connect(":memory:", autocommit=False)
>>> cx.in_transaction
True
>>> cx.execute("commit")
<sqlite3.Cursor object at 0x10d2b3f50>
>>> cx.in_transaction
False
>>> cx.commit()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
sqlite3.OperationalError: cannot commit - no transaction is active
```

Reply:
Yes in the mean time I updated my comment. It seems we agree =)

Reply:
> Maybe this deserves a comment for future readers?

Yeah, a comment might be useful here.